### PR TITLE
backend/feat: #618-Adding-driver-silent-mode

### PR DESCRIPTION
**Pull request by umangj-28**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

## Description
Currently drivers only have Online and Offline Mode, and if driver is using another app while Online, on closing the ride popup driver is considered Offline.
Introducing Silent Mode-> where driver can be switched to "Silent:" mode when he is using other apps and still will be considered in driver pool.

### Additional Changes

- [x] This PR modifies the database schema (database migration added)
- [ ] This PR modifies dhall configs/environment variables
- [ ] This PR contains API breaking changes

## How did you test it?
tested locally

## Checklist

- [x] I formatted the code and addressed linter errors `./dev/format-all-files.sh`
- [x] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/API/UI/Driver.hs [hunk 1]
```diff
@@ -144,8 +146,8 @@ createDriver admin = withFlowHandlerAPI . DDriver.createDriver admin
 getInformation :: Id SP.Person -> FlowHandler DDriver.DriverInformationRes
 getInformation = withFlowHandlerAPI . DDriver.getInformation
 
-setActivity :: Id SP.Person -> Bool -> FlowHandler APISuccess
-setActivity personId = withFlowHandlerAPI . DDriver.setActivity personId
+setActivity :: Id SP.Person -> Bool -> Maybe DI.DriverMode -> FlowHandler APISuccess
+setActivity personId mode = withFlowHandlerAPI . DDriver.setActivity personId mode
```
Comment (piyushKumar-1) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/API/UI/Driver.hs:150]: it should be `isActive` not `mode`.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Allocator/Jobs/SendSearchRequestToDrivers/Handle/Internal/DriverPool.hs [hunk 2]
```diff
@@ -142,10 +143,13 @@ prepareDriverPoolBatch driverPoolCfg searchReq batchNum = withLogTag ("BatchNum-
           poolBatch = sortedPart <> randomPart
           driversFromRestCount = take (driversCount - length poolBatch) (restRandom <> restSorted) -- taking rest of drivers if poolBatch length is less then driverCount requried.
           finalPoolBatch = poolBatch <> driversFromRestCount
+      let (silentDrivers, activeDrivers) = DL.partition ((== Just DriverInfo.SILENT) . (.driverPoolResult.mode)) finalPoolBatch
+      let finalPoolBatchWithSilentDrivers = activeDrivers <> silentDrivers
       logDebug $ "IntelligentDriverPool - SortedDriversCount " <> show (length sortedPart)
       logDebug $ "IntelligentDriverPool - RandomizedDriversCount " <> show (length randomPart)
       logDebug $ "IntelligentDriverPool - finalPoolBatch " <> show (length finalPoolBatch)
```
Comment (piyushKumar-1) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Allocator/Jobs/SendSearchRequestToDrivers/Handle/Internal/DriverPool.hs:150]: this log can be removed.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/Person.hs [hunk 3]
```diff
@@ -507,6 +508,7 @@ getNearestDrivers mbVariant LatLong {..} radiusMeters merchantId onlyNotOnRide m
       person ^. PersonRole ==. val Person.DRIVER
         &&. person ^. PersonMerchantId ==. val (toKey merchantId)
         &&. driverInfo ^. DriverInformationActive
+        &&. driverInfo ^. DriverInformationMode ==. val (Just DriverInfo.SILENT) ||. driverInfo ^. DriverInformationMode ==. val (Just DriverInfo.ONLINE) ||. Esq.isNothing (driverInfo ^. DriverInformationMode)
```
Comment (piyushKumar-1) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/Person.hs:511]: can be written with `whenJust_` instead ?

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Driver.hs [hunk 4]
```diff
@@ -406,21 +411,21 @@ getInformation personId = do
...
+setActivity personId isActive mode = do
   _ <- QPerson.findById personId >>= fromMaybeM (PersonNotFound personId.getId)
   let driverId = cast personId
-  when isActive $ do
+  when (isActive || (isJust mode && (mode == Just DriverInfo.SILENT || mode == Just DriverInfo.ONLINE))) $ do
     driverInfo <- QDriverInformation.findById driverId >>= fromMaybeM DriverInfoNotFound
     unless driverInfo.enabled $ throwError DriverAccountDisabled
     unless (not driverInfo.blocked) $ throwError DriverAccountBlocked
-  QDriverInformation.updateActivity driverId isActive
+  QDriverInformation.updateActivity driverId isActive mode
```
Comment (piyushKumar-1) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/UI/Driver.hs:422]: can we keep the isActive as True and don't use it at all in pool calculation ? otherwise depreciating it will become difficult.

**Review by piyushKumar-1 — APPROVED** [review]

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Cancel.hs [hunk 5]
```diff
@@ -93,9 +93,9 @@ cancel transporterId _ req = do
     whenJust mbRide $ \ride -> do
       QRide.updateStatus ride.id SRide.CANCELLED
       driverInfo <- QDI.findById (cast ride.driverId) >>= fromMaybeM (PersonNotFound ride.driverId.getId)
-      if driverInfo.active
-        then QDFS.updateStatus ride.driverId DDFS.ACTIVE
-        else QDFS.updateStatus ride.driverId DDFS.IDLE
+      Esq.runTransaction $
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Cancel.hs:96]: why runTransaction inside a runTransaction?

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Confirm.hs [hunk 6]
```diff
@@ -321,9 +322,9 @@ cancelBooking booking mbDriver transporter = do
     whenJust mbRide $ \ride -> do
       QRide.updateStatus ride.id DRide.CANCELLED
       driverInfo <- QDI.findById (cast ride.driverId) >>= fromMaybeM (PersonNotFound ride.driverId.getId)
-      if driverInfo.active
-        then QDFS.updateStatus ride.driverId DDFS.ACTIVE
-        else QDFS.updateStatus ride.driverId DDFS.IDLE
+      Esq.runTransaction $
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Domain/Action/Beckn/Confirm.hs:325]: same

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Allocator/Jobs/SendSearchRequestToDrivers/Handle/Internal/DriverPool.hs [hunk 7]
```diff
@@ -132,7 +134,6 @@ prepareDriverPoolBatch driverPoolCfg searchReq batchNum = withLogTag ("BatchNum-
 
         makeRandomDriverPool batchSize onlyNewDrivers = take batchSize <$> randomizeAndLimitSelection onlyNewDrivers
 
-    takeDriversUsingPoolPercentage :: (MonadFlow m) => ([a], [a]) -> Int -> DriverIntelligentPoolConfig -> m [a]
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Allocator/Jobs/SendSearchRequestToDrivers/Handle/Internal/DriverPool.hs:135]: why did you remove this?

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Allocator/Jobs/SendSearchRequestToDrivers/Handle/Internal/DriverPool.hs [hunk 8]
```diff
@@ -142,10 +143,13 @@ prepareDriverPoolBatch driverPoolCfg searchReq batchNum = withLogTag ("BatchNum-
           poolBatch = sortedPart <> randomPart
           driversFromRestCount = take (driversCount - length poolBatch) (restRandom <> restSorted) -- taking rest of drivers if poolBatch length is less then driverCount requried.
           finalPoolBatch = poolBatch <> driversFromRestCount
+      let (silentDrivers, activeDrivers) = DL.partition ((== Just DriverInfo.SILENT) . (.driverPoolResult.mode)) finalPoolBatch
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Allocator/Jobs/SendSearchRequestToDrivers/Handle/Internal/DriverPool.hs:146]: This logic should be before we split the on basis of random and intelligent because there is no benefit of doing it after doing the split. We anyway send the request to all drivers in finalPoolBatch.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/Person.hs [hunk 9]
```diff
@@ -507,6 +508,7 @@ getNearestDrivers mbVariant LatLong {..} radiusMeters merchantId onlyNotOnRide m
       person ^. PersonRole ==. val Person.DRIVER
         &&. person ^. PersonMerchantId ==. val (toKey merchantId)
         &&. driverInfo ^. DriverInformationActive
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/Person.hs:510]: Active should not be checked if `DriverInformationMode` is Just.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/Person.hs [hunk 10]
```diff
@@ -633,6 +637,7 @@ getNearestDriversCurrentlyOnRide mbVariant LatLong {..} radiusMeters merchantId
       personInfo ^. PersonRole ==. val Person.DRIVER
         &&. personInfo ^. PersonMerchantId ==. val (toKey merchantId)
         &&. driverInfo ^. DriverInformationActive
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/Person.hs:639]: same here.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/API/UI/Driver.hs (hunk 1 above)
Comment (umangj-28) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/API/UI/Driver.hs:150]: isActive is being passed on as before

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/Person.hs (hunk 3 above)
Comment (umangj-28) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/Person.hs:511]: Tried it, whenJust_ is not getting applied on a table column, its used for values i guess

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Allocator/Jobs/SendSearchRequestToDrivers/Handle/Internal/DriverPool.hs (hunk 7 above)
Comment (umangj-28) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Allocator/Jobs/SendSearchRequestToDrivers/Handle/Internal/DriverPool.hs:135]: because I need DriverMode param from DriverPoolResult type inside this function, but here in declaration we change the obj type from  DriverPoolWithActualDistResult to **[a]**

Comment (umangj-28) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/Allocator/Jobs/SendSearchRequestToDrivers/Handle/Internal/DriverPool.hs:135]: adding it back as logic will shift before sortingtype check
